### PR TITLE
`define-pool` relative to the project root

### PR DESCRIPTION
**Pull request by fosskers**
As mentioned in the docs, `:base` assumes that the `data/` directory for assets is in the same level than the source file being compiled. `:root` relaxes this, making the lookup relative to the project root as detected through `+app-system+`.

However, usage of `:root` requires that the user set `+app-system+` as follows:
```lisp
(eval-when (:compile-toplevel :load-toplevel :execute)
  (setf +app-system+ "save-the-farm"))
```
Otherwise, `+app-system+` will still be set to `trial` (the default) during the first macro-expansion of a project-local `define-pool`, and paths won't resolve correctly. I avoided using `(data-root)` as it caches its result, and would become permanently set to `trial` otherwise.

### Semantics

- `<nothing>`: Current behaviour, no change. `=> src/data/`
- `:base "foo"`: Current behaviour, no change. `=> src/data/foo/`
- `:base "/absolute/path/"`: Current behaviour, no change. `=> /absolute/path/data/`
- `:root t`: This addition. `=> <project>/data/`
- `:root t :base "foo"`: This addition. `=> <project>/foo/data/`

Although now that I look, (2) and (5) seem to be in conflict with respect to the position of `data/`. I suppose it's an option question about the primacy of the `data` folder; whether it's supposed to come first, last, or be baked-in at all.

### Alternate Approaches

- Allow relative pathnames to be passed to `:base` (e.g. `..`). Currently not as possible due to how `pathname-utils` elides `..`.
- (Breaking) Have `:base` always operate from the project root, not the directory of the compiling file.
- Avoid referencing `+app-system+` and detect the project root some other way.

### CODE REVIEW

File: asset-pool.lisp [hunk 1]
```diff
@@ -76,17 +76,23 @@
...
     `(eval-when (:compile-toplevel :load-toplevel :execute)
-       (cond ((find-pool ',name)
-              (reinitialize-instance (find-pool ',name) ,@initargs))
-             (T
-              (setf (find-pool ',name) (make-instance 'pool :name ',name ,@initargs :base ,path))))
+       (let ((pool (find-pool ',name)))
+         (cond ((and pool (equal ,path (base pool)))
+                (reinitialize-instance pool ,@initargs))
+               (T
+                (setf (find-pool ',name) (make-instance 'pool :name ',name ,@initargs :base ,path)))))
```
Comment (fosskers) [inline asset-pool.lisp:95]: I left a repeated `find-pool` call here; using the `pool` binding here directly won't work due to `setf` accessor semantics, right?

File: asset-pool.lisp [hunk 2]
```diff
@@ -76,17 +76,23 @@
...
+                    (pathname-utils:subdirectory path "data"))))
     (remf initargs :base)
+    (remf initargs :root)
     `(eval-when (:compile-toplevel :load-toplevel :execute)
-       (cond ((find-pool ',name)
-              (reinitialize-instance (find-pool ',name) ,@initargs))
-             (T
-              (setf (find-pool ',name) (make-instance 'pool :name ',name ,@initargs :base ,path))))
+       (let ((pool (find-pool ',name)))
+         (cond ((and pool (equal ,path (base pool)))
```
Comment (fosskers) [inline asset-pool.lisp:92]: Only reinitialize if the `path` didn't change. Is `equal` correct here?

**Comment by Shinmera** [comment]
What do you mean by this?

> Allow relative pathnames to be passed to :base (e.g. ..). Currently not as possible due to how pathname-utils elides ...

`define-pool` internally does
```lisp
(pathname-utils:merge-pathnames* #p"../b/c" (pathname-utils:subdirectory #p "a/" "data"))
; => #P"a/data/../b/c"
```
which works as expected, no?

**Comment by fosskers** [comment]
A few samples, at least executed locally:
```lisp
(pathname-utils:merge-pathnames*
 #p".."
 (pathname-utils:subdirectory "/home/colin/code/common-lisp/foo/src/bar.lisp" "data"))
=> #P"/home/colin/code/common-lisp/foo/src/data/."
```

**Comment by Shinmera** [comment]
#p".." is a file called .., not a directory, so that's just you getting pathname syntax wrong. 

As for why it complains, it seems your file just doesn't exist or you got the path wrong somehow. The following in workbench.lisp works as expected:

```lisp
(define-pool workbench
  :base "../../trial-assets/data/")

(define-asset (workbench test) image
    #p "box-diffuse.png")
```

Assuming your structure has the directories trial and trial-assets side-by-side.

**Comment by fosskers** [comment]
Back on `master` for a moment to verify:
paths
In this case, `src/` and `data/` are side-by-side in the project root, and while it can establish the pool, it can't find the assets.

**Comment by fosskers** [comment]
I just noticed that the pool's `ASSETS` slot has two entries, the lemon looking like:
lemon

Although `(launch)` nonetheless fails, claiming it can't find the assets.

**Comment by Shinmera** [comment]
That file test is very odd indeed. Are there symlinks involved somehow?

**Comment by fosskers** [comment]
> Are there symlinks involved somehow?

Nope they're all real. Did you mean that the first `file-exists-p` should have yielded non-NIL? Does it on your machine? Is there god-forbid a Mac/Linux difference here?

**Comment by Shinmera** [comment]
Anyway, I pushed a potential fix that tries to reduce the path, try it with that.

**Comment by fosskers** [comment]
```lisp
(define-pool farm :base #p"../../data/")
```
This now works, provided I have the latest `pathname-utils` pulled. Thanks! Let's go with this for now.

**Pull request closed without merging**